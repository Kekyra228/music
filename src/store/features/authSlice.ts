import { authorize } from "@/app/api/authApi";
import { fetchTokens, fetchUser } from "@/app/api/logonApi";
import { StaredUserType } from "@/types/types";
import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

type SigninFormTypes = {
  email: string;
  password: string;
};
type SignUpFormTypes = {
  email: string;
  password: string;
  username: string;
};

export const createUser = createAsyncThunk(
  "user/createUser",
  async ({ email, password, username }: SignUpFormTypes) => {
    const user = await authorize({ email, password, username });
    return user;
  }
);

export const getUser = createAsyncThunk(
  "user/getUser",
  async ({ email, password }: SigninFormTypes) => {
    const user = await fetchUser({ email, password });
    return user;
  }
);

export const getTokens = createAsyncThunk(
  "user/getTokens",
  async ({ email, password }: SigninFormTypes) => {
    const tokens = await fetchTokens({ email, password });
    return tokens;
  }
);

type AuthStateType = {
  newUser: null | StaredUserType;
  user: null | StaredUserType;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
};

const initialState: AuthStateType = {
  newUser: null,
  user: null,
  tokens: {
    access: null,
    refresh: null,
  },
};

const authSlice = createSlice({
  name: "auth",
  initialState,
  reducers: {
    logout: (state) => {
      state.user = null;
      state.tokens.access = null;
      state.tokens.refresh = null;
    },
  },
  extraReducers(builder) {
    builder
      .addCase(
        getUser.fulfilled,
        (state, action: PayloadAction<StaredUserType>) => {
          state.user = action.payload;
          // const user = localStorage.setItem("user", JSON.stringify(user));
        }
      )
      .addCase(
        getTokens.fulfilled,
        (
          state,
          action: PayloadAction<{
            access: string | null;
            refresh: string | null;
          }>
        ) => {
          state.tokens.access = action.payload.access;
          state.tokens.refresh = action.payload.refresh;
        }
      );
  },
});
export const { logout } = authSlice.actions;
export const authReducer = authSlice.reducer;
