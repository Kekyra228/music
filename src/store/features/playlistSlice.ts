import { fetchFavoriteTracks, addLike, removeLike } from "@/app/api/userApi";
import { TrackType } from "@/types/types";
import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

export const getFavoriteTracks = createAsyncThunk(
  "playlist/getFavoriteTracks",
  async (access: string) => {
    const favoriteTracks = await fetchFavoriteTracks(access);
    return favoriteTracks;
  }
);
export const addLikeInTrack = createAsyncThunk(
  "playlist/getFavoriteTracks",
  async (id: number) => {
    const likedTrack = await addLike(id);
    return likedTrack;
  }
);

export const removeLikeInTrack = createAsyncThunk(
  "playlist/getFavoriteTracks",
  async (id: number) => {
    const dislikedTrack = await removeLike(id);
    return dislikedTrack;
  }
);

type PlaylistStateType = {
  isPlaying: boolean;
  currentTrack: TrackType | null;
  tracks: TrackType[];
  isShuffled: boolean;
  shuffledPlaylist: TrackType[];
  searchFilter: {
    author: string[];
    genre: string[];
    orderSorting: string;
    searchString: string;
  };
  filtredPlaylist: TrackType[];
  likedTracks: number[];
};

const initialState: PlaylistStateType = {
  isPlaying: false,
  currentTrack: null,
  tracks: [],
  isShuffled: false,
  shuffledPlaylist: [],
  searchFilter: {
    author: [],
    genre: [],
    orderSorting: "По умолчанию",
    searchString: "",
  },
  filtredPlaylist: [],
  likedTracks: [],
};

const playlistSlice = createSlice({
  name: "playlist",
  initialState,
  reducers: {
    setCurrentTrack: (
      state,
      action: PayloadAction<{ currentTrack: TrackType; tracks: TrackType[] }>
    ) => {
      state.currentTrack = action.payload.currentTrack;
      state.tracks = action.payload.tracks;
      state.shuffledPlaylist = [...action.payload.tracks].sort(
        () => 0.5 - Math.random()
      );
    },
    nextTrack: (state) => {
      const playlist = state.isShuffled ? state.shuffledPlaylist : state.tracks;
      const currentIndex = playlist.findIndex(
        (track) => track.id === state.currentTrack?.id
      );
      const nextIndex = currentIndex + 1;
      if (nextIndex >= playlist.length) {
        state.currentTrack = playlist[0];
      } else {
        state.currentTrack = playlist[nextIndex];
      }
    },
    prevTrack: (state) => {
      const playlist = state.isShuffled ? state.shuffledPlaylist : state.tracks;
      const currentIndex = playlist.findIndex(
        (track) => track.id === state.currentTrack?.id
      );
      const prevIndex = currentIndex - 1;
      state.currentTrack = playlist[prevIndex];
    },
    setShuffle: (state) => {
      state.isShuffled = !state.isShuffled;
    },
    setIsPlaying: (state) => {
      state.isPlaying = !state.isPlaying;
    },
    setPlaylist: (state, action: PayloadAction<{ tracks: TrackType[] }>) => {
      state.tracks = action.payload.tracks;
      state.filtredPlaylist = action.payload.tracks;
    },
    setLikedTracks: (state, action: PayloadAction<number[]>) => {
      state.likedTracks = action.payload;
    },
    setFilter: (
      state,
      action: PayloadAction<{
        author?: string[];
        genre?: string[];
        orderSorting?: string;
        searchString?: string;
      }>
    ) => {
      state.searchFilter = {
        author: action.payload.author || state.searchFilter.author,
        genre: action.payload.genre || state.searchFilter.genre,
        orderSorting:
          action.payload.orderSorting || state.searchFilter.orderSorting,
        searchString:
          action.payload.searchString || state.searchFilter.searchString,
      };
      const filterTracks = [...state.tracks].filter((track) => {
        const searchString =
          // track.author &&
          track.name
            .toLocaleLowerCase()
            .includes(state.searchFilter.searchString.toLocaleLowerCase());
        const hasAuthorFilter =
          state.searchFilter.author.length > 0
            ? state.searchFilter.author.includes(track.author)
            : true;
        const hasGenreFilter =
          state.searchFilter.genre.length > 0
            ? state.searchFilter.genre.includes(track.genre)
            : true;
        return searchString && hasAuthorFilter && hasGenreFilter;
      });
      switch (state.searchFilter.orderSorting) {
        case "Сначала новые":
          filterTracks.sort(
            (a, b) =>
              new Date(b.release_date).getTime() -
              new Date(a.release_date).getTime()
          );
          break;
        case "Сначала старые":
          filterTracks.sort(
            (a, b) =>
              new Date(a.release_date).getTime() -
              new Date(b.release_date).getTime()
          );
          break;
        default:
          break;
      }
      state.filtredPlaylist = filterTracks;
    },
    likeTrack: (state, action: PayloadAction<{ id: number }>) => {
      state.likedTracks.push(action.payload.id);
    },
    dislike: (state, action: PayloadAction<{ id: number }>) => {
      state.likedTracks = state.likedTracks.filter(
        (el) => el !== action.payload.id
      );
    },
  },
});

export const {
  setCurrentTrack,
  nextTrack,
  prevTrack,
  setShuffle,
  setIsPlaying,
  setFilter,
  setPlaylist,
  likeTrack,
  dislike,
  setLikedTracks,
} = playlistSlice.actions;
export const playlistReducer = playlistSlice.reducer;
