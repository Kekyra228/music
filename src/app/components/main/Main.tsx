"use client";
import { TrackType } from "@/types/types";
import SongList from "../songsList/SongsList";
import { useAppDispatch, useAppSelector } from "@/hooks/store";
import { setPlaylist } from "@/store/features/playlistSlice";
import { useCallback, useEffect, useState } from "react";

type Props = {
  tracks: TrackType[];
};

export default function Main({ tracks }: Props) {
  const [isLoading, setIsLoading] = useState(true);
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(setPlaylist({ tracks }));
    setIsLoading(false);
  }, [tracks, dispatch]);

  const filtredTracks = useAppSelector(
    (store) => store.playlist.filtredPlaylist
  );
  return (
    <>
      {" "}
      {isLoading ? "Загрузка треков..." : <SongList tracks={filtredTracks} />}
    </>
  );
}
