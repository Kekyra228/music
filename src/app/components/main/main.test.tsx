import React from "react";
import renderer from "react-test-renderer";
import Main from "./Main";
import { Provider } from "react-redux";
import "@testing-library/jest-dom/extend-expect";
import configureStore from "redux-mock-store";
import { initialState } from "@/store/features/playlistSlice";

describe("Компонент Main", () => {
  const mockStore = configureStore([]);
  let store = mockStore({ playlist: initialState });

  test("компонент рендерится", () => {
    const component = renderer.create(
      <Provider store={store}>
        <Main tracks={[]} />
      </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
