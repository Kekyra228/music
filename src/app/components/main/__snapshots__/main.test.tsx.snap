// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент Main компонент рендерится 1`] = `
[
  " ",
  "Загрузка треков...",
]
`;
